### PR TITLE
Added AdaptiveCards Template WinRT Component

### PR DESCRIPTION
**Pull request by jwoo-msft**
# Related Issue
Added AdaptiveCards Template WinRT Component

# Description
Added C# WinRT component of templating library, and C++ WinRT console app sample that demonstrate the usage.
WinUI Visualizer changes to follow.

# How Verified
* Sample app usage
* Existing template samples were used.

### CODE REVIEW

File: source/dotnet/Library/CppConsoleApp/readme.txt [hunk 1]
```diff
@@ -0,0 +1,30 @@
+========================================================================
```
Comment (paulcam206) [inline source/dotnet/Library/CppConsoleApp/readme.txt:1]: don't need this file

File: source/dotnet/NuGet.config [hunk 2]
```diff
@@ -4,6 +4,7 @@
     <!-- remove any machine-wide sources with <clear/> -->
     <clear />
     <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+    <add key="local" value="packages" />
```
Comment (paulcam206) [inline source/dotnet/NuGet.config:7]: probably don't want to check this change in

File: source/dotnet/Library/CppConsoleApp/pch.cpp [hunk 3]
```diff
@@ -0,0 +1 @@
+﻿#include "pch.h"
```
Comment (paulcam206) [inline source/dotnet/Library/CppConsoleApp/pch.cpp:1]: missing copyright banners here and below

File: source/dotnet/Library/AdaptiveCards.Templating.CSharp.WinRT/AdaptiveCards.Template.csproj [hunk 4]
```diff
@@ -0,0 +1,24 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0-windows10.0.19041.0</TargetFramework>
+    <Platforms>x64</Platforms>
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards.Templating.CSharp.WinRT/AdaptiveCards.Template.csproj:5]: this should be extended to include `arm64` and `x86`

File: source/dotnet/Library/AdaptiveCards.Templating.CSharp.WinRT/AdaptiveCardsTemplate.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        public AdaptiveCardTemplate(String jsonTemplate)
+        {
+            adaptiveCardTemplate = new AdaptiveCards.Templating.AdaptiveCardTemplate(jsonTemplate);
+        }
+
+        public string Expand(String rootData)
+        {
+            return adaptiveCardTemplate.Expand(rootData);
+        }
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards.Templating.CSharp.WinRT/AdaptiveCardsTemplate.cs:17]: missing docstrings

File: source/dotnet/AdaptiveCards.Visualizer.Wpf.sln [hunk 6]
```diff
@@ -1,7 +1,7 @@
 ﻿
```
Comment (paulcam206) [inline source/dotnet/AdaptiveCards.Visualizer.Wpf.sln:1]: is it possible to get some tests in place?

File: source/dotnet/Library/CppConsoleApp/CppConsoleApp.vcxproj [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      <Platform>Win32</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Debug|x64">
+      <Configuration>Debug</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Release|x64">
+      <Configuration>Release</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
```
Comment (paulcam206) [inline source/dotnet/Library/CppConsoleApp/CppConsoleApp.vcxproj:33]: missing `arm64` and `x86` (`Win32`)

File: source/dotnet/Library/CppConsoleApp/CppConsoleApp.vcxproj [hunk 8]
```diff
@@ -0,0 +1,137 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (paulcam206) [inline source/dotnet/Library/CppConsoleApp/CppConsoleApp.vcxproj:1]: even though this binary doesn't ship, it'll probably save a fair amount of pain if you turn on compliance flags

https://github.com/microsoft/AdaptiveCards/blob/d3c25221fdebaa8631e96e132e5329b71e5ccab1/source/uwp/winui3/AdaptiveCardsObjectModel/AdaptiveCardsObjectModel.vcxproj#L134-L136

https://github.com/microsoft/AdaptiveCards/blob/d3c25221fdebaa8631e96e132e5329b71e5ccab1/source/uwp/winui3/AdaptiveCardsObjectModel/AdaptiveCardsObjectModel.vcxproj#L107-L108

(I'm probably missing some here)

File: source/dotnet/AdaptiveCards.Visualizer.Wpf.sln (hunk 6 above)
Comment (paulcam206) [inline source/dotnet/AdaptiveCards.Visualizer.Wpf.sln:1]: do we need pipeline changes here, or is that going to be a subsequent PR?

**Review by paulcam206 — CHANGES_REQUESTED** [review]

File: source/dotnet/AdaptiveCards.Visualizer.Wpf.sln (hunk 6 above)
Comment (jwoo-msft) [reply source/dotnet/AdaptiveCards.Visualizer.Wpf.sln:1]: we will need pipeline change; setting pack command with versioning and publishing path and etc.

Comment (jwoo-msft) [reply source/dotnet/AdaptiveCards.Visualizer.Wpf.sln:1]: I'm currently adding templating function to winui visualizer; as I work on it; i'm going to add unit testing as well.

File: source/dotnet/Library/CppConsoleApp/CppConsoleApp.vcxproj [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+  <Import Project="..\..\packages\Microsoft.Windows.CppWinRT.2.0.240111.5\build\native\Microsoft.Windows.CppWinRT.props" Condition="Exists('..\..\packages\Microsoft.Windows.CppWinRT.2.0.240111.5\build\native\Microsoft.Windows.CppWinRT.props')" />
+  <PropertyGroup Label="Globals">
+    <CppWinRTOptimized>true</CppWinRTOptimized>
+    <CppWinRTRootNamespaceAutoMerge>true</CppWinRTRootNamespaceAutoMerge>
+    <CppWinRTGenerateWindowsMetadata>true</CppWinRTGenerateWindowsMetadata>
+    <MinimalCoreWin>true</MinimalCoreWin>
+    <VCProjectVersion>15.0</VCProjectVersion>
+    <ProjectGuid>{ed459eec-7211-403e-8c08-78a43a05a7be}</ProjectGuid>
+    <Keyword>Win32Proj</Keyword>
+    <RootNamespace>CppConsoleApp</RootNamespace>
```
Comment (almedina-ms) [inline source/dotnet/Library/CppConsoleApp/CppConsoleApp.vcxproj:12]: can you rename this file to something more explicit about what does this project do? If it's an app to test that templating work on a cpp app you can rename it to TemplatingTestAppCpp or something similar

File: source/dotnet/Library/AdaptiveCards.Templating.CSharp.WinRT/AdaptiveCardsTemplate.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /// <summary>
+    /// AdaptiveCards template windows runtime component library
+    /// </summary>
+    public sealed class AdaptiveCardTemplate
+    {
+        /// <summary>
+        /// Given a JSON template, create a AdaptiveCardTemplate object that can be used to expand the template
+        /// </summary>
+        /// <param name="jsonTemplate"></param>
```
Comment (almedina-ms) [inline source/dotnet/Library/AdaptiveCards.Templating.CSharp.WinRT/AdaptiveCardsTemplate.cs:15]: empty param description

File: source/dotnet/Library/AdaptiveCards.Templating.CSharp.WinRT/AdaptiveCardsTemplate.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        public AdaptiveCardTemplate(String jsonTemplate)
+        {
+            adaptiveCardTemplate = new AdaptiveCards.Templating.AdaptiveCardTemplate(jsonTemplate);
+        }
+
+        /// <summary>
+        /// Expands the template with the given data
+        /// </summary>
+        /// <param name="rootData"></param>
```
Comment (almedina-ms) [inline source/dotnet/Library/AdaptiveCards.Templating.CSharp.WinRT/AdaptiveCardsTemplate.cs:24]: empty param description and empty result

File: source/dotnet/Library/CppConsoleApp/main.cpp [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    winrt::hstring input{
+        L"{\"type\": \"AdaptiveCard\",\"body\": [{\"type\": \"TextBlock\",\"text\": \"Hello, ${hi}!\"}]}"};
+
+    winrt::AdaptiveCards::Template::AdaptiveCardTemplate adaptiveCardTemplate{ input };
+
+    winrt::hstring data{ L"{\"hi\": \"world\"}" };
+    auto result = adaptiveCardTemplate.Expand(data);
+
+    std::cout << "Expanded: " << winrt::to_string(result) << std::endl;
+};
```
Comment (almedina-ms) [inline source/dotnet/Library/CppConsoleApp/main.cpp:24]: nit: missing return

File: source/dotnet/Library/CppConsoleApp/main.cpp [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+using namespace winrt;
+using namespace Windows::Foundation;
+
+// Sample Test App for AdaptiveCards.Template project which is a C#/WinRT component
+int main()
+{
+    init_apartment();
+
+    winrt::hstring input{
```
Comment (almedina-ms) [inline source/dotnet/Library/CppConsoleApp/main.cpp:15]: if you already added the "using namespace winrt;" directive, you should be able to remove the "winrt::" prefix from all the spots you're using it in this file

File: source/dotnet/Library/CppConsoleApp/packages.config [hunk 14]
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="AdaptiveCards.Template" version="1.0.0" targetFramework="native" />
```
Comment (almedina-ms) [inline source/dotnet/Library/CppConsoleApp/packages.config:3]: is this line presuming the existence of the AdaptiveCards.Template package that supports C++ code?

**Review by almedina-ms — CHANGES_REQUESTED** [review]

File: source/dotnet/Library/CppConsoleApp/packages.config (hunk 14 above)
Comment (jwoo-msft) [reply source/dotnet/Library/CppConsoleApp/packages.config:3]: nuget package is generated by the new template project.

File: source/dotnet/Library/CppConsoleApp/main.cpp [hunk 15]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "iostream"
```
Comment (paulcam206) [inline source/dotnet/Library/CppConsoleApp/main.cpp:5]: nitpick I missed earlier:

```suggestion
#include <iostream>
```

File: source/dotnet/Library/CppConsoleApp/pch.h [hunk 16]
```diff
@@ -0,0 +1,4 @@
+#pragma once
```
Comment (paulcam206) [inline source/dotnet/Library/CppConsoleApp/pch.h:1]: missing banner

**Review by paulcam206 — APPROVED** [review]

File: source/dotnet/NuGet.config (hunk 2 above)
Comment (jwoo-msft) [reply source/dotnet/NuGet.config:7]: we need to drop console app cause we are using locally produced nuget, but we run nuget restore task before the build; it won't work. it's catch 22.

File: source/dotnet/AdaptiveCards.sln [hunk 17]
```diff
@@ -561,6 +653,118 @@ Global
 		{9E7E5953-A5B4-4867-9A06-046A754AFAC1}.Release|x64.Build.0 = Release|Any CPU
 		{9E7E5953-A5B4-4867-9A06-046A754AFAC1}.Release|x86.ActiveCfg = Release|Any CPU
 		{9E7E5953-A5B4-4867-9A06-046A754AFAC1}.Release|x86.Build.0 = Release|Any CPU
+		{1D73A371-9637-4E08-B78F-F6BCC5872007}.Ad-Hoc|Any CPU.ActiveCfg = Debug|x64
+		{1D73A371-9637-4E08-B78F-F6BCC5872007}.Ad-Hoc|Any CPU.Build.0 = Debug|x64
+		{1D73A371-9637-4E08-B78F-F6BCC5872007}.Ad-Hoc|ARM.ActiveCfg = Debug|x64
+		{1D73A371-9637-4E08-B78F-F6BCC5872007}.Ad-Hoc|ARM.Build.0 = Debug|x64
+		{1D73A371-9637-4E08-B78F-F6BCC5872007}.Ad-Hoc|ARM64.ActiveCfg = Debug|arm64
+		{1D73A371-9637-4E08-B78F-F6BCC5872007}.Ad-Hoc|ARM64.Build.0 = Debug|arm64
+		{1D73A371-9637-4E08-B78F-F6BCC5872007}.Ad-Hoc|iPhone.ActiveCfg = Debug|x64
```
[the remark this answers is not in the document]
Comment (jwoo-msft) [reply source/dotnet/AdaptiveCards.sln]: this isn't only project that gets these options if you target "any cpu"

File: source/dotnet/Library/AdaptiveCards.Templating.CSharp.WinRT/AdaptiveCardsTemplate.cs [hunk 18]
```diff
@@ -0,0 +1,34 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// AdaptiveCards template windows runtime component library
+namespace AdaptiveCards.Template
+{
+    /// <summary>
+    /// AdaptiveCards template windows runtime component library
+    /// </summary>
+    public sealed class AdaptiveCardTemplate
```
[the remark this answers is not in the document]
Comment (jwoo-msft) [reply source/dotnet/Library/AdaptiveCards.Templating.CSharp.WinRT/AdaptiveCardsTemplate.cs:10]: this is basically the midl for winrt component.

**Review by almedina-ms — APPROVED** [review]

**Review by paulcam206 — APPROVED** [review]

File: source/dotnet/NuGet.config (hunk 2 above)
Comment (jwoo-msft) [reply source/dotnet/NuGet.config]: removed the locals

**Pull request merged**